server/client.go

```diff
@@ -755,28 +755,41 @@ func (c *client) processErr(errStr string) {
 }
 
 // Password pattern matcher.
-var passPat = regexp.MustCompile(`"?\s*pass\S*?"?[:=]\s*"?(([^"])*)`)
+var passPat = regexp.MustCompile(`"?\s*pass\S*?"?\s*[:=]\s*"?(([^",\r\n}])*)`)
 
 // This will remove any notion of passwords from trace messages
 // for logging.
 func removePassFromTrace(arg []byte) []byte {
-
 	if !bytes.Contains(arg, []byte("pass")) {
 		return arg
 	}
-	m := passPat.FindAllSubmatch(arg, -1)
+	// Take a copy of the connect proto just for the trace message.
+	a := make([]byte, len(arg))
+	copy(a, arg)
+
+	m := passPat.FindAllSubmatchIndex(a, -1)
 	if len(m) == 0 {
-		return arg
+		return a
 	}
 
-	for _, match := range m {
-		if len(match) != 3 {
+	j := 0
+	redactedPass := []byte("[REDACTED]")
+	redactedPassLen := len(redactedPass)
+	for _, i := range m {
+		if len(i) < 6 {
 			continue
 		}
-		arg = bytes.Replace(arg, match[1], []byte("[REDACTED]"), 1)
+		// Need to shift the index in case password appears more than once,
+		// since the substrings are being removed in place.
+		start := i[2] + j
+		end := i[5] + j
+		passLen := end - start
+		j += redactedPassLen - passLen
```

Comment:
Maybe would be more straightforward to just use JSON to filter the entries here too (more overhead though)

https://github.com/wallyqs/gnatsd/commit/89e397f08f1937393d87d900fe60efe4f969807d#diff-853eb184ac73cf9597d7833f6b89e9c9R782

Reply:
I had thought of that but thought it would be expensive. But that may be the right approach, we are moving users aways from u/p. @kozlovic WDYT?

Reply:
It is a CONNECT so this is not as fast-path as messages. I would favor safety and correctness here and not focus so much on performance. Also this is when trace is enabled, so... :-)

Reply:
Great points. @wallyqs lets look into the JSON parse path. I was cautious that the unmarshal/marshal step would change other things besides the password that may be beneficial to a trace, so let's watch for that.

Reply:
Hmm.. Does it keep all the keys though at least?

Reply:
If some keys are repeated, then only the last one would be part of the result, for example:

```json
{"user":"s3cr3t","pass":"AAA","pass":"BBB","password":"CCC", "foo":"hi", "foo":"hello"}
```

...in the trace logs would become:

```json
{"foo":"hello","pass":"[REDACTED]","password":"[REDACTED]","user":"s3cr3t"}`
```

Reply:
To me this is a non issue because it would mean that the original object is malformed (having duplicate keys), which if coming from a server or client would not happen.

Reply:
Unless a new or custom client with a bug.